File: Backtracking/79_WordSearch.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        path = set()
        directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]

        def dfs(r, c, i):
            if board[r][c] != word[i]:
                return False
            if i == len(word) - 1:
                return True

            path.add((r, c))
            res = False
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and (nr, nc) not in path:
                    if dfs(nr, nc, i + 1):
                        res = True
                        break
            path.remove((r, c))
            return res

        for r in range(rows):
            for c in range(cols):
                if dfs(r, c, 0):
                    return True
        return False

File: Java Solutions/79_Word Search.java
class Solution {
    private char[][] board;
    private int ROWS;
    private int COLS;
    private String word;

    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.ROWS = board.length;
        this.COLS = board[0].length;
        this.word = word;

        for (int r = 0; r < this.ROWS; r++) {
            for (int c = 0; c < this.COLS; c++) {
                if (this.dfs(r, c, 0)) return true;
            }
        }
        return false;
    }

    private boolean dfs(int r, int c, int i) {
        if (i == this.word.length()) return true;

        if (
            r < 0 || r >= this.ROWS ||
            c < 0 || c >= this.COLS ||
            this.board[r][c] != word.charAt(i)
        ) return false;

        this.board[r][c] = '#';
        boolean res = false;

        int[] rDirections = {0, 1, 0, -1};
        int[] cDirections = {1, 0, -1, 0};
        for (int d = 0; d < 4; d++) {
            res = this.dfs(r + rDirections[d], c + cDirections[d], i + 1);
            if (res) break;
        }

        this.board[r][c] = word.charAt(i);
        return res;
    }
}